adapters/synacormedia/synacormedia.go

```diff
@@ -0,0 +1,184 @@
+
+	// create JSON Request Body
+	request.Imp = validImps
+	request.Ext, err = json.Marshal(re)
+	if err != nil {
+		return nil, append(errs, err)
+	}
+
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, append(errs, err)
+	}
+
+	// set Request Headers
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+
+	// create Request Uri
+	reqUri, err := a.buildEndpointURL(firstExtImp)
+	if err != nil {
+		return nil, append(errs, err)
+	}
+
+	return &adapters.RequestData{
+		Method:  http.MethodPost,
+		Uri:     reqUri,
+		Body:    reqJSON,
+		Headers: headers,
+	}, errs
+}
+
+// Builds enpoint url based on adapter-specific pub settings from imp.ext
+func (adapter *SynacorMediaAdapter) buildEndpointURL(params *openrtb_ext.ExtImpSynacormedia) (string, error) {
+	return macros.ResolveMacros(adapter.EndpointTemplate, macros.EndpointTemplateParams{Host: params.SeatId})
+}
+
+func getExtImpObj(imp *openrtb.Imp) (*openrtb_ext.ExtImpSynacormedia, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var synacormediaExt openrtb_ext.ExtImpSynacormedia
+	if err := json.Unmarshal(bidderExt.Bidder, &synacormediaExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	return &synacormediaExt, nil
+}
+
+// MakeBids make the bids for the bid response.
+func (a *SynacorMediaAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	const errorMessage string = "Unexpected status code: %d. Run with request.debug = 1 for more info"
+	switch {
+	case response.StatusCode == http.StatusNoContent:
+		return nil, nil
+	case response.StatusCode == http.StatusBadRequest:
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf(errorMessage, response.StatusCode),
+		}}
+	case response.StatusCode != http.StatusOK:
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf(errorMessage, response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb.BidResponse
+
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
+
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &sb.Bid[i],
+				BidType: getMediaTypeForImp(sb.Bid[i].ImpID, internalRequest.Imp),
+			})
+		}
+	}
+	return bidResponse, nil
+}
+
+func getMediaTypeForImp(impId string, imps []openrtb.Imp) openrtb_ext.BidType {
+	mediaType := openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impId {
+			if imp.Banner == nil && imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeVideo
+			}
+			return mediaType
+		}
+	}
+	return mediaType
+}
```

Comment:
Function `getMediaTypeForImp(impId string, imps []openrtb.Imp)` might return `openrtb_ext.BidTypeBanner` even if `imp.Banner == nil` and `imp.Video == nil`. Other adapters overcome this issue by filtering out non-Banner and non-Video types somewhere inside `MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo)`. I believe either filtering out non-Banner and non-Video imps, or modifying this function to return an error or something so we don't append this bid to the `bidResponse` array in `MakeBids` may solve this issue.
```
173 func getMediaTypeForImp(impId string, imps []openrtb.Imp) openrtb_ext.BidType {
174     mediaType := openrtb_ext.BidTypeBanner
175     for _, imp := range imps {
176         if imp.ID == impId {
177             if imp.Banner == nil && imp.Video != nil {
178                 mediaType = openrtb_ext.BidTypeVideo
179             }
180             return mediaType  //<- may return openrtb_ext.BidTypeBanner on an imp with imp.Banner==nil && imp.Video==nil element
181         }
182     }
183     return mediaType  //<- may return openrtb_ext.BidTypeBanner even if imp wasn't found in the imp list
184 }
synacormedia.go
```

Reply:
Added filter.

```go
	// filter out non banner and non video imps
	var validRequestImps []openrtb.Imp
	for _, imp := range internalRequest.Imp {
		if imp.Banner != nil || imp.Video != nil {
			validRequestImps = append(validRequestImps, imp)
		}
	}
```

Reply:
I don't think this for loop is exactly filtering out non banner and non video imps because if `getMediaTypeForImp(sb.Bid[i].ImpID, validRequestImps)` doesn't find a match in `validRequestImps` it will still assign the value `openrtb_ext.BidTypeBanner` to something that probably isn't. Do you think a better approach is this?
```
138 func (a *SynacorMediaAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
139     const errorMessage string = "Unexpected status code: %d. Run with request.debug = 1 for more info"
140     switch {
141     case response.StatusCode == http.StatusNoContent:
142         return nil, nil
143     case response.StatusCode == http.StatusBadRequest:
144         return nil, []error{&errortypes.BadInput{
145             Message: fmt.Sprintf(errorMessage, response.StatusCode),
146         }}
147     case response.StatusCode != http.StatusOK:
148         return nil, []error{&errortypes.BadServerResponse{
149             Message: fmt.Sprintf(errorMessage, response.StatusCode),
150         }}
151     }
152
153     var bidResp openrtb.BidResponse
154
155     if err := json.Unmarshal(response.Body, &bidResp); err != nil {
156         return nil, []error{err}
157     }
158
159     bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
160
161  -  // filter out non banner and non video imps
162  -  var validRequestImps []openrtb.Imp
163  -  for _, imp := range internalRequest.Imp {
164  -      if imp.Banner != nil || imp.Video != nil {
165  -          validRequestImps = append(validRequestImps, imp)
166  -      }
167  -  }
168  -
169     for _, sb := range bidResp.SeatBid {
170         for i := range sb.Bid {
     +          mediaType = getMediaTypeForImp(sb.Bid[i].ImpID, validRequestImps)
     +          if mediaType != openrtb_ext.BidTypeBanner && mediaType != openrtb_ext.BidTypeVideo {
     +              continue
     +          }
171             bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
172                 Bid:     &sb.Bid[i],
173  -              BidType: getMediaTypeForImp(sb.Bid[i].ImpID, validRequestImps),
173                 BidType: mediaType,
174             })
175         }
176     }
177     return bidResponse, nil
178 }
179
180 func getMediaTypeForImp(impId string, imps []openrtb.Imp) openrtb_ext.BidType {
181     mediaType := openrtb_ext.BidTypeBanner
182     for _, imp := range imps {
183         if imp.ID == impId {
184  -          if imp.Banner == nil && imp.Video != nil {
185  -              mediaType = openrtb_ext.BidTypeVideo
     +          if imp.Banner != nil {
     +              break
     +          }
     +          if imp.Video != nil {
     +              mediaType = openrtb_ext.BidTypeVideo
     +              break
     +          }
     +          if imp.Native != nil {
     +              mediaType = openrtb_ext.BidTypeNative
     +              break
     +          }
     +          if imp.Audio != nil {
     +              mediaType = openrtb_ext.BidTypeAudio
     +              break
186             }
187  -          return mediaType
188         }
189     }
190     return mediaType
191 }
adapters/synacormedia/synacormedia.go
```